File: code/Java/src/main/java/MarsRoverController.java
import App.CommandInterpreter;
import App.MarsRoverEngine;
import Commands.ICommand;
import Model.Position;

import java.util.ArrayList;

public class MarsRoverController implements IProcessMessages {
    private MarsRoverEngine marsRoverEngine;
    private CommandInterpreter commandInterpreter;
    private ISendNotificationBus marsRoverServiceWriter;

    public MarsRoverController() {
        this.marsRoverEngine = new MarsRoverEngine();
        commandInterpreter = new CommandInterpreter();
    }

    public void writesTo(ISendNotificationBus marsRoverServiceBus) {

        this.marsRoverServiceWriter = marsRoverServiceBus;
    }

    public void readsFrom(IReadMessages marsRoverServiceBus) {

        marsRoverServiceBus.callBack(this);
    }

    public void process(String messageReceived) {
        ArrayList<ICommand> commands = commandInterpreter.translate(messageReceived);
        marsRoverEngine.execute(commands);
        Position finalPosition = marsRoverEngine.getPosition();
        marsRoverServiceWriter.NotifyExecution(finalPosition.toString());
    }
}

File: code/CSharp/ConnascentMarsRover/Source/Infrastructure/MarsRoverController.cs
using Source.App;
using Source.Infrastructure.Bus;
using Source.Model;

namespace Source.Infrastructure
{
    public class MarsRoverController : IProcessMessages
    {
        private readonly MarsRoverEngine _marsRoverEngine;
        private readonly CommandInterpreter _commandInterpreter;
        private ISendNotificationBus _marsRoverServiceWriter;

        public MarsRoverController()
        {
            _marsRoverEngine = new MarsRoverEngine();
            _commandInterpreter = new CommandInterpreter();
        }

        public void WritesTo(ISendNotificationBus marsRoverServiceBus)
        {
            _marsRoverServiceWriter = marsRoverServiceBus;
        }

        public void ReadsFrom(IReadMessages marsRoverServiceBus)
        {
            marsRoverServiceBus.CallBack(this);
        }

        public void Process(string messageReceived)
        {
            var commands = _commandInterpreter.Translate(messageReceived);
            _marsRoverEngine.Execute(commands);
            var finalPosition = _marsRoverEngine.GetPosition();
            _marsRoverServiceWriter.NotifyExecution(finalPosition.ToString());
        }
    }
}